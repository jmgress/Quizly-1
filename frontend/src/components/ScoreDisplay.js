import React from 'react';

const ScoreDisplay = ({ results, questions, onRestart }) => {
  const getScoreMessage = () => {
    const percentage = results.score_percentage;
    
    if (percentage >= 90) {
      return "üèÜ Excellent! You're a quiz master!";
    } else if (percentage >= 70) {
      return "üéâ Great job! Well done!";
    } else if (percentage >= 50) {
      return "üëç Good effort! Keep practicing!";
    } else {
      return "üìö Keep studying and try again!";
    }
  };

  const getQuestionById = (questionId) => {
    return questions.find(q => q.id === questionId);
  };

  const getOptionText = (question, optionId) => {
    const option = question?.options.find(opt => opt.id === optionId);
    return option?.text || 'Unknown';
  };

  return (
    <div className="card">
      <div className="score-display">
        <h2>üéØ Quiz Complete!</h2>
        
        <div className="score-number">
          {results.correct_answers}/{results.total_questions}
        </div>
        
        <div className="score-message">
          {Math.round(results.score_percentage)}% Correct
        </div>
        
        <p>{getScoreMessage()}</p>
        
        <div style={{ marginTop: '30px', textAlign: 'left' }}>
          <h3>üìã Review Your Answers:</h3>
          
          {results.answers.map((answer, index) => {
            const question = getQuestionById(answer.question_id);
            if (!question) return null;
            
            return (
              <div 
                key={answer.question_id} 
                style={{ 
                  marginBottom: '20px', 
                  padding: '15px', 
                  backgroundColor: answer.is_correct ? '#d4edda' : '#f8d7da',
                  border: `1px solid ${answer.is_correct ? '#c3e6cb' : '#f5c6cb'}`,
                  borderRadius: '8px'
                }}
              >
                <p style={{ fontWeight: 'bold', marginBottom: '8px' }}>
                  Q{index + 1}: {question.text}
                </p>
                
                <p style={{ marginBottom: '4px' }}>
                  <strong>Your answer:</strong> {getOptionText(question, answer.selected_answer)}
                  {answer.is_correct ? ' ‚úÖ' : ' ‚ùå'}
                </p>
                
                {!answer.is_correct && (
                  <p style={{ marginBottom: '0', color: '#155724' }}>
                    <strong>Correct answer:</strong> {getOptionText(question, answer.correct_answer)}
                  </p>
                )}
              </div>
            );
          })}
        </div>
        
        <div style={{ marginTop: '30px' }}>
          <button className="button" onClick={onRestart}>
            Take Another Quiz
          </button>
        </div>
      </div>
    </div>
  );
};

export default ScoreDisplay;