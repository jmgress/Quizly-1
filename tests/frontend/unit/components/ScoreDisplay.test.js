import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import ScoreDisplay from '../../../../../frontend/src/components/ScoreDisplay';

// Mock data for testing
const mockQuestions = [
  {
    id: 1,
    text: 'What is the capital of France?',
    options: [
      { id: 'a', text: 'London' },
      { id: 'b', text: 'Berlin' },
      { id: 'c', text: 'Paris' },
      { id: 'd', text: 'Madrid' }
    ],
    correct_answer: 'c'
  },
  {
    id: 2,
    text: 'What is 2 + 2?',
    options: [
      { id: 'a', text: '3' },
      { id: 'b', text: '4' },
      { id: 'c', text: '5' },
      { id: 'd', text: '6' }
    ],
    correct_answer: 'b'
  }
];

const mockResultsHighScore = {
  correct_answers: 2,
  total_questions: 2,
  score_percentage: 100,
  answers: [
    {
      question_id: 1,
      selected_answer: 'c',
      correct_answer: 'c',
      is_correct: true
    },
    {
      question_id: 2,
      selected_answer: 'b',
      correct_answer: 'b',
      is_correct: true
    }
  ]
};

const mockResultsLowScore = {
  correct_answers: 1,
  total_questions: 2,
  score_percentage: 50,
  answers: [
    {
      question_id: 1,
      selected_answer: 'a',
      correct_answer: 'c',
      is_correct: false
    },
    {
      question_id: 2,
      selected_answer: 'b',
      correct_answer: 'b',
      is_correct: true
    }
  ]
};

const mockResultsVeryLowScore = {
  correct_answers: 0,
  total_questions: 2,
  score_percentage: 0,
  answers: [
    {
      question_id: 1,
      selected_answer: 'a',
      correct_answer: 'c',
      is_correct: false
    },
    {
      question_id: 2,
      selected_answer: 'a',
      correct_answer: 'b',
      is_correct: false
    }
  ]
};

describe('ScoreDisplay Component', () => {
  let mockOnRestart;

  beforeEach(() => {
    mockOnRestart = jest.fn();
  });

  test('displays the correct score based on props - high score', () => {
    render(
      <ScoreDisplay
        results={mockResultsHighScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );

    // Check score display
    expect(screen.getByText('2/2')).toBeInTheDocument();
    expect(screen.getByText('100% Correct')).toBeInTheDocument();
    expect(screen.getByText("🏆 Excellent! You're a quiz master!")).toBeInTheDocument();
  });

  test('displays the correct score based on props - medium score', () => {
    render(
      <ScoreDisplay
        results={mockResultsLowScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );

    // Check score display
    expect(screen.getByText('1/2')).toBeInTheDocument();
    expect(screen.getByText('50% Correct')).toBeInTheDocument();
    expect(screen.getByText('👍 Good effort! Keep practicing!')).toBeInTheDocument();
  });

  test('displays the correct score based on props - low score', () => {
    render(
      <ScoreDisplay
        results={mockResultsVeryLowScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );

    // Check score display
    expect(screen.getByText('0/2')).toBeInTheDocument();
    expect(screen.getByText('0% Correct')).toBeInTheDocument();
    expect(screen.getByText('📚 Keep studying and try again!')).toBeInTheDocument();
  });

  test('calls onRestart when restart button is clicked', () => {
    render(
      <ScoreDisplay
        results={mockResultsHighScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );

    const restartButton = screen.getByText('Take Another Quiz');
    fireEvent.click(restartButton);

    expect(mockOnRestart).toHaveBeenCalledTimes(1);
  });

  test('displays correct answer review for correct answers', () => {
    render(
      <ScoreDisplay
        results={mockResultsHighScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );

    // Check that questions are displayed
    expect(screen.getByText(/Q1:/)).toBeInTheDocument();
    expect(screen.getByText(/What is the capital of France?/)).toBeInTheDocument();
    expect(screen.getByText(/Q2:/)).toBeInTheDocument();
    expect(screen.getByText(/What is 2 \+ 2?/)).toBeInTheDocument();
    
    // Check that "Your answer:" text appears
    const yourAnswerTexts = screen.getAllByText(/Your answer:/);
    expect(yourAnswerTexts).toHaveLength(2);
  });

  test('displays correct answer review for incorrect answers', () => {
    render(
      <ScoreDisplay
        results={mockResultsLowScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );

    // Check that questions are displayed
    expect(screen.getByText(/Q1:/)).toBeInTheDocument();
    expect(screen.getByText(/What is the capital of France?/)).toBeInTheDocument();
    expect(screen.getByText(/Q2:/)).toBeInTheDocument();
    expect(screen.getByText(/What is 2 \+ 2?/)).toBeInTheDocument();
    
    // Check that "Correct answer:" text appears for incorrect answers
    expect(screen.getByText(/Correct answer:/)).toBeInTheDocument();
    
    // Check that both "Your answer:" appear
    const yourAnswerTexts = screen.getAllByText(/Your answer:/);
    expect(yourAnswerTexts).toHaveLength(2);
  });

  test('renders quiz complete title', () => {
    render(
      <ScoreDisplay
        results={mockResultsHighScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );

    expect(screen.getByText('🎯 Quiz Complete!')).toBeInTheDocument();
    expect(screen.getByText('📋 Review Your Answers:')).toBeInTheDocument();
  });

  test('matches snapshot for different score values - high score', () => {
    const { container } = render(
      <ScoreDisplay
        results={mockResultsHighScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('matches snapshot for different score values - low score', () => {
    const { container } = render(
      <ScoreDisplay
        results={mockResultsVeryLowScore}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );
    expect(container.firstChild).toMatchSnapshot();
  });

  test('displays different score messages based on percentage', () => {
    const mockResults90 = { ...mockResultsHighScore, score_percentage: 95 };
    const mockResults75 = { ...mockResultsHighScore, score_percentage: 75 };
    const mockResults55 = { ...mockResultsHighScore, score_percentage: 55 };
    const mockResults25 = { ...mockResultsHighScore, score_percentage: 25 };

    // Test excellent message (≥90%)
    const { rerender } = render(
      <ScoreDisplay
        results={mockResults90}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );
    expect(screen.getByText("🏆 Excellent! You're a quiz master!")).toBeInTheDocument();

    // Test great job message (≥70%)
    rerender(
      <ScoreDisplay
        results={mockResults75}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );
    expect(screen.getByText('🎉 Great job! Well done!')).toBeInTheDocument();

    // Test good effort message (≥50%)
    rerender(
      <ScoreDisplay
        results={mockResults55}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );
    expect(screen.getByText('👍 Good effort! Keep practicing!')).toBeInTheDocument();

    // Test keep studying message (<50%)
    rerender(
      <ScoreDisplay
        results={mockResults25}
        questions={mockQuestions}
        onRestart={mockOnRestart}
      />
    );
    expect(screen.getByText('📚 Keep studying and try again!')).toBeInTheDocument();
  });
});