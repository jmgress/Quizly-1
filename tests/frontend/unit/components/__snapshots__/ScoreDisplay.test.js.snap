// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ScoreDisplay Component matches snapshot for different score values - high score 1`] = `
<div
  class="card"
>
  <div
    class="score-display"
  >
    <h2>
      🎯 Quiz Complete!
    </h2>
    <div
      class="score-number"
    >
      2
      /
      2
    </div>
    <div
      class="score-message"
    >
      100
      % Correct
    </div>
    <p>
      🏆 Excellent! You're a quiz master!
    </p>
    <div
      style="margin-top: 30px; text-align: left;"
    >
      <h3>
        📋 Review Your Answers:
      </h3>
      <div
        style="margin-bottom: 20px; padding: 15px; background-color: rgb(212, 237, 218); border: 1px solid rgb(195, 230, 203); border-radius: 8px;"
      >
        <p
          style="font-weight: bold; margin-bottom: 8px;"
        >
          Q
          1
          : 
          What is the capital of France?
        </p>
        <p
          style="margin-bottom: 4px;"
        >
          <strong>
            Your answer:
          </strong>
           
          Paris
           ✅
        </p>
      </div>
      <div
        style="margin-bottom: 20px; padding: 15px; background-color: rgb(212, 237, 218); border: 1px solid rgb(195, 230, 203); border-radius: 8px;"
      >
        <p
          style="font-weight: bold; margin-bottom: 8px;"
        >
          Q
          2
          : 
          What is 2 + 2?
        </p>
        <p
          style="margin-bottom: 4px;"
        >
          <strong>
            Your answer:
          </strong>
           
          4
           ✅
        </p>
      </div>
    </div>
    <div
      style="margin-top: 30px;"
    >
      <button
        class="button"
      >
        Take Another Quiz
      </button>
    </div>
  </div>
</div>
`;

exports[`ScoreDisplay Component matches snapshot for different score values - low score 1`] = `
<div
  class="card"
>
  <div
    class="score-display"
  >
    <h2>
      🎯 Quiz Complete!
    </h2>
    <div
      class="score-number"
    >
      0
      /
      2
    </div>
    <div
      class="score-message"
    >
      0
      % Correct
    </div>
    <p>
      📚 Keep studying and try again!
    </p>
    <div
      style="margin-top: 30px; text-align: left;"
    >
      <h3>
        📋 Review Your Answers:
      </h3>
      <div
        style="margin-bottom: 20px; padding: 15px; background-color: rgb(248, 215, 218); border: 1px solid rgb(245, 198, 203); border-radius: 8px;"
      >
        <p
          style="font-weight: bold; margin-bottom: 8px;"
        >
          Q
          1
          : 
          What is the capital of France?
        </p>
        <p
          style="margin-bottom: 4px;"
        >
          <strong>
            Your answer:
          </strong>
           
          London
           ❌
        </p>
        <p
          style="margin-bottom: 0px; color: rgb(21, 87, 36);"
        >
          <strong>
            Correct answer:
          </strong>
           
          Paris
        </p>
      </div>
      <div
        style="margin-bottom: 20px; padding: 15px; background-color: rgb(248, 215, 218); border: 1px solid rgb(245, 198, 203); border-radius: 8px;"
      >
        <p
          style="font-weight: bold; margin-bottom: 8px;"
        >
          Q
          2
          : 
          What is 2 + 2?
        </p>
        <p
          style="margin-bottom: 4px;"
        >
          <strong>
            Your answer:
          </strong>
           
          3
           ❌
        </p>
        <p
          style="margin-bottom: 0px; color: rgb(21, 87, 36);"
        >
          <strong>
            Correct answer:
          </strong>
           
          4
        </p>
      </div>
    </div>
    <div
      style="margin-top: 30px;"
    >
      <button
        class="button"
      >
        Take Another Quiz
      </button>
    </div>
  </div>
</div>
`;
